Semihost: open sys call doesn't handle "rb+" mode

mostafa-salmaan opened the pull request:
### According to the ISO/IEC_9899_1999, section:7.19.5.3

- Opening file with read mode ('r' as the first character in the mode argument) fails if the file does not exist.
In case of `r+`, `r+b`, `rb+` Semihost system call "open" implemented in `open.c` does not handle those update modes ("+" is update mode) only handles `wb+` and `ab+` which results in 2 issues:
1. Opening a non-existing file using mentioned modes creates the file which is not the correct behavior according to the standard.
2. Opening a file using the mentioned modes then write into it, it appends as it is `ab+` mode because it is the "else" case.

### Test Case (Reproduces the Issue)
```c
#include <stdio.h>
#include <stdlib.h>

int main() {
    // Open a non-existent file to update (read+write), should fail according to ISO C
    FILE *res = fopen("this_file_does_not_exist.dat", "rb+");

    if (res == NULL) {
        printf("[PASS] fopen() failed as expected for non-existent file\n");
    } else {
        printf("[FAIL] fopen() unexpectedly succeeded (file should not be created)\n");
        fclose(res);
        return 1;
    }

    return 0;
}

```
This test fails before the fix and passes after the fix — verifying that `fopen("rb+")` does not create a file if it does not exist, as per the ISO standard.

keith-packard commented on semihost/open.c:
```diff
@@ -71,6 +71,8 @@ open(const char *pathname, int flags, ...)
 			semiflags = SH_OPEN_W_PLUS_B;	/* 'wb+' */
 		else if (flags & O_APPEND)
 			semiflags = SH_OPEN_A_PLUS_B;	/* 'ab+' */
+		else if (flags & O_CREAT)
```
This isn't right either -- using 'ab+' will put the initial file position at the end of the file and cause new writes
to append to the file. I couldn't find a mode that will open the file for reading and writing, creating it if necessary, without placing the file in append mode. Maybe this requires a sequence of two operations? First attempt to use 'rb+', and if that fails, use 'wb+'?

keith-packard commented on semihost/open.c:
```diff
@@ -69,8 +69,10 @@ open(const char *pathname, int flags, ...)
 		if (flags & O_TRUNC)
 			semiflags = SH_OPEN_W_PLUS_B;	/* 'wb+' */
-		else
+		else if (flags & O_APPEND)
```
Hrm. None of the modes which create the file are checking for O_CREAT. That seems wrong.

mostafa-salmaan commented:
Hi Keith,

I've updated the fix to remove the `O_CREAT` check from the default case. The logic now only checks for `O_TRUNC` and `O_APPEND`, falling back to `rb+` otherwise:
```c
default:
    if (flags & O_TRUNC)
        semiflags = SH_OPEN_W_PLUS_B;  /* 'wb+' */
    else if (flags & O_APPEND)
        semiflags = SH_OPEN_A_PLUS_B;  /* 'ab+' */
    else
        semiflags = SH_OPEN_R_PLUS_B;  /* 'rb+' */
    break;

```
This change ensures that modes like `"r+", "rb+", "r+b"` (which currently fall back to 'ab+') now behave correctly by failing if the file doesn't exist, per ISO/IEC 9899:1999 (section 7.19.5.3).

I've also added a test case in the PR description that reproduces the issue by attempting to open a non-existent file with `rb+`. It correctly fails before the fix and passes afterward.

Additionally, I noticed the CI is currently failing, and I'm looking into the logs to identify the cause. Will update shortly once I have more details.

keith-packard commented:
Yeah, from my reading it's not possible to map all of the possible combinations of open flags into the limited set of semihost open modes. If you want to make it work better, it's going to take multiple semihost operations to implement some of the possible flag combinations.

mostafa-salmaan commented:
Thanks, @keith — you're absolutely right.

I agree that the limited set of semihost open modes makes full compatibility tricky, especially when trying to support all flag combinations. For now, my intent was just to correct the specific case where `"r+", "rb+", "r+b"`, were falling through to `ab+`, leading to non-conformant behavior (e.g., creating files that shouldn't exist or appending instead of updating).

This change doesn't aim to fully model all `O_*` combinations, but rather to improve correctness in the most common cases without adding complexity. That said, I understand that a complete solution would likely require multiple semihost calls (e.g., checking file existence before deciding how to open it).

Would you be okay with merging this fix as an incremental improvement, given that it addresses the specific misbehavior around `"r+", "rb+", "r+b"`, without trying to cover all possible flag combinations?

keith-packard commented:
I'm not sure it matters as you're just changing the set of bugs here. If you want to actually fix this function to work correctly, I'd say go for it. But, it's just the semihosting code, which is not expected to be used in production applications, so it's not of the highest priority for correctness in the library.

mostafa-salmaan commented:
This change improves the semihosting implementation to support more accurate behavior during standard compliance verification. While semihosting is not intended for production use, it can be a valuable part of a qualification framework, where the focus is on observing expected and standard-conforming behavior from the DUT.

In my case, I tested this change using Picolibc with the ARC GNU Tools toolchain, and found that it resolved previously failed issues. This demonstrates that the enhanced semihosting behavior can be beneficial when using Picolibc in verification environments.
